### PR TITLE
Implemented the Hyperbolic Sine Function for matrix 

### PR DESCRIPTION
**Pull request by ameyachawla99**
Hi , I am Ameya Chawla currently pursuing my B.Tech in Computer Science and Engineering .
I read about your organization in GSOC 2021 and looking forward to work with your organization .

Added the implementation of method "Hyperbolic Sine Function"(sinhm)
as mentioned in the comparison document

As discussed in #4942 i have added code to  my code to matfuncs.py
https://docs.cupy.dev/en/stable/reference/comparison.html

### CODE REVIEW

**Comment by emcastillo** [comment]
please add tests

**Comment by ameyachawla99** [comment]
Where should i add test cases?

**Comment by emcastillo** [comment]
In tests/cupyx_tests/scipy_tests/linalg_tests folder.

File: cupyx/scipy/linalg/matfuncs.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from numpy import exp
+
+
+def sinhm(arr):
+    """Returns hyperbolic sine of the given square matrix.
+Args: arr: square matrix .
+Returns: Hyperbolic sine of given square matrix as input.
```
Comment (emcastillo) [inline cupyx/scipy/linalg/matfuncs.py:7]: Please, format the docstring with identation and line breaks

File: cupyx/scipy/linalg/matfuncs.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+from numpy import exp
+
+
+def sinhm(arr):
+    """Returns hyperbolic sine of the given square matrix.
+Args: arr: square matrix .
+Returns: Hyperbolic sine of given square matrix as input.
+    """
+    return 0.5*(exp(arr)-exp(-1*arr))
```
Comment (emcastillo) [inline cupyx/scipy/linalg/matfuncs.py:9]:
```suggestion
    return 0.5 * (exp(arr) - exp(-1 * arr))
```
We leave a space between operands.

File: cupyx/scipy/linalg/matfuncs.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+from numpy import exp
+
+
+def sinhm(arr):
+    """Returns hyperbolic sine of the given square matrix.
+
+Args:.
```
Comment (emcastillo) [inline cupyx/scipy/linalg/matfuncs.py:7]: Please look other cupy functions docstring and use the same format:

https://github.com/cupy/cupy/blob/master/cupyx/scipy/linalg/decomp_lu.py#L13-L35

**Comment by emcastillo** [comment]
you can install flake8 locally with `pip install flake8` and check the file with `flake8 cupyx/scipy/linalg/matfuncs.py` so you don't need to push everytime to check the changes.

File: cupyx/scipy/linalg/matfuncs.py (hunk 3 above)
Comment (ameyachawla99) [reply cupyx/scipy/linalg/matfuncs.py:7]: I will make the necessary changes.

**Comment by ameyachawla99** [comment]
I have made changes in matfuncs.py , please can your guide me how should I make the test file?

**Comment by emcastillo** [comment]
look for other tests, ideally you should add tests for edge cases and common ones, empty arrays, invalid arguments, etc.

**Comment by kmaehashi** [comment]
Tests reside in the directory with `test_` prefixed for the corresponding module name.
For example, tests code for `cupyx/scipy/linalg` will be in `tests/cupyx_tests/scipy_tests/linalg_tests`. 

To start reviewing, please write (at least) one most simple test case to verify your implementation is correct, and confirm that the test pass **in your local machine** (this is important!) before pushing it to the pull-request.

**Comment by ameyachawla99** [comment]
I am also adding the conditions for edge cases as if wrong input given by user should produce error .

**Comment by ameyachawla99** [comment]
I have added the edge cases and test file .

File: cupyx/scipy/linalg/matfuncs.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    Returns hyperbolic sine of the given square matrix.
+    Args:.
+        arr(cupy.ndarray) : Square matrix whose hyperbolic .
+        sine has to be calculated .
+    Returns: .
+        (cupy.ndarray):  Hyperbolic sine of given square matrix as input.
+    """
+    arr = numpy.asanyarray(arr)
+
+    "Checking whether the input is a 2D matrix or not"
```
Comment (kmaehashi) [inline cupyx/scipy/linalg/matfuncs.py:15]: For single-line comments use `#` instead.

File: cupyx/scipy/linalg/matfuncs.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import numpy
+
+
+def sinhm(arr):
+    """Hyperbolic Sine
+    Returns hyperbolic sine of the given square matrix.
+    Args:.
```
Comment (kmaehashi) [inline cupyx/scipy/linalg/matfuncs.py:7]: Please follow the doc style of other functions.

File: tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
+from matfuncs import sinhm
+import numpy
+import unittest
+
+
+class TestMatFuncsNan(unittest.TestCase):
+
+    "Testing for nan values in matrix"
```
Comment (kmaehashi) [inline tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:8]: Comments for functions must be written inside the function.

File: tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+from matfuncs import sinhm
+import numpy
+import unittest
+
+
+class TestMatFuncsNan(unittest.TestCase):
```
Comment (kmaehashi) [inline tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: It seems your test cases only cover anomaly cases, and using NumPy array as an input.

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py (hunk 7 above)
Comment (ameyachawla99) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: I was converting array in input as NumPy array to use functions defined for NumPy array like shape.

Comment (ameyachawla99) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: Which other test cases should i also add?

Comment (ameyachawla99) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: I have edited the files .
Please guide me for any further changes .

Comment (asi1024) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: @ameyachawla99 Did you check if this test passes locally?

Comment (ameyachawla99) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: Yes all 5 tests were passes 
i have added the screenshot of terminal .
Screenshot from 2021-03-27 12-25-38

Comment (asi1024) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: Why `from matfuncs import sinhm` works? `matfuncs.py` is not placed in test directory.

Comment (asi1024) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: Anyway, this test code looks to just call NumPy routines. Could you use `cupy.ndarray` for the input?

Comment (ameyachawla99) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: > Why `from matfuncs import sinhm` works? `matfuncs.py` is not placed in test directory.

While i was testing i placed matfuncs in same directory .

Comment (ameyachawla99) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: > call NumPy routines

Sure i will update it .

Comment (ameyachawla99) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: > Anyway, this test code looks to just call NumPy routines. Could you use `cupy.ndarray` for the input?

But then changes has to be made in both files as error will be caused of cupy.ndarray as not of same datatype of numpy.ndarray

Comment (ameyachawla99) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: @asi1024 i have made the changes in both files so that they can use cupy.ndarray instead of numpy.ndarray

Comment (asi1024) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: Could you add an alias `cupyx.scipy.linalg.sinhm` and use it in the test?

Comment (asi1024) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: SciPy's documentation says that `scipy.linalg.sinhm` uses **matrix exponentials**, but your implementation uses element-wise exponential.  https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.sinhm.htm.

Comment (ameyachawla99) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: Then i will first define expm which is used 

> SciPy's documentation says that `scipy.linalg.sinhm` uses **matrix exponentials**, but your implementation uses element-wise exponential. https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.sinhm.htm.

Then i will first define expm

Comment (kmaehashi) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: Oh, sorry I overlooked that `cupyx.scipy.sparse.linalg.expm` is not implemented yet... I should have noticed it and comment that in the first review round. You can either try implementing it in this PR or in a separate PR.

Comment (ameyachawla99) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:6]: Actually it is my fault i did not understood the function properly
